File: Baekjoon/김경훈/Q1012.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');
const T = input.shift();
let count; 
let need = 0;

const directX = [1, -1, 0, 0]; // 좌, 우 방문을 위한 배열 선언
const directY = [0, 0, 1, -1]; // 상, 하 방문을 위한 배열 선언

for(let i = 0; i < T; i++) { // T만큼 반복한다
    count = 0; // 배추가 심어진 땅 카운트
    const [N, M, K] = input.shift().split(' ').map((el) => parseInt(el));
    
    let graph = Array.from(Array(N + 1), () => Array(M + 1).fill(0)); // N x M 배열의 인접행렬 생성
    let limitCount = 0; // graph 채우기 위한 변수
    
    while(limitCount < K) { // input에서 shift해가며 인접행렬 채우기
        const [front, back] = input.shift().split(' ');
        graph[front][back] = 1;
        limitCount += 1;
    }

    const DFS = (a, b) => { // DFS 구현
        if(a >= 0 && a < N && b >= 0 && b < M && graph[a][b] === 1) { // 지도 밖을 벗어나지 않기 위한 조건문 && 배추가 있다면
            graph[a][b] = 0; // 방문 처리
            count += 1; // 배추 개수 count
            for(let k = 0; k < 4; k++) { // 땅 상, 하, 좌, 우 차례 방문
                DFS(a + directX[k], b + directY[k]);
            }
        }
    }

    // 그래프 탐색
    for (let i = 0; i < N; i++) {
        for (let j = 0; j < N; j++) { // 인접행렬 순회
            if (graph[i][j] === 1) { // 배추가 있다면
                DFS(i, j); // DFS 호출 

                // 한 번의 순회가 끝나고 나면
                if(count > 0) { // 배추가 있는 땅이 여러개라면
                    count = 0; // 카운트 0 초기화
                    need += 1; // 지렁이 개수 +
                }
            }
        }
    }
    
    console.log(need) // 필요한 지렁이 수 출력
    need = 0;

}

File: Baekjoon/유석환/Q1012.swift
let T = Int(readLine()!)! // 테스트 케이스의 개수

for _ in 0..<T {
    let input = readLine()!.split(separator: " ").map { Int(String($0))! }
    let M = input[0] // 배추밭의 가로길이 M
    let N = input[1] // 배추밭의 세로길이 N
    let K = input[2] // 배추의 개수 K

    var graph = [[Int]](repeating: [Int](repeating: 0, count: M), count: N) // 그래프 생성
    for _ in 0..<K {
        let cabbage = readLine()!.split(separator: " ").map { Int(String($0))! }
        graph[cabbage[1]][cabbage[0]] = 1 // 배추의 위치 저장
    }
    
    var result = 0 // 배추흰지렁이 수
    
    func dfs(_ x: Int, _ y: Int) {
        if x < 0 || x >= M || y < 0 || y >= N { // index out of range 방지
            return
        }
        
        if graph[y][x] == 1 { // 배추가 있으면
            graph[y][x] = 0 // 방문처리
            
            dfs(x + 1, y) // 상하좌우 노드 재귀 호출
            dfs(x - 1, y)
            dfs(x, y + 1)
            dfs(x, y - 1)
        }
    }
    
    for y in 0..<N { // 배추밭 순회
        for x in 0..<M {
            if graph[y][x] == 1 { // 배추가 있으면
                dfs(x, y) // dfs 호출
                result += 1 // 배추흰지렁이 수 증가
            }
        }
    }
    
    print(result) // 배추흰지렁이 수 출력
}
